src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -78,80 +83,129 @@ public void onRefreshSettings(Settings settings) {
+            ArrayList<FsChannelImmutableReader> foundTranslogs = new ArrayList<>();
+            for (Path location : locations) {
+                try (DirectoryStream<Path> stream = Files.newDirectoryStream(location, TRANSLOG_FILE_PREFIX + "[0-9]*")) {
+                    for (Path file : stream) {
+                        final long id = parseIdFromFileName(file);
+                        if (id < 0) {
+                            throw new TranslogException(shardId, "failed to parse id from file name matching pattern " + file);
+                        }
+                        idGenerator = Math.max(idGenerator, id + 1);
+                        final ChannelReference raf = new InternalChannelReference(id, location.resolve(getPath(id)), StandardOpenOption.READ);
+                        foundTranslogs.add(new FsChannelImmutableReader(id, raf, raf.channel().size(), FsChannelReader.UNKNOWN));
+                        logger.debug("found local translog with id [{}]", id);
+                    }
+                }
+            }
+            CollectionUtil.timSort(foundTranslogs, new Comparator<FsChannelReader>() {
+                @Override
+                public int compare(FsChannelReader o1, FsChannelReader o2) {
+                    return Long.compare(o1.translogId(), o2.translogId());
+                }
+            });
+            uncommittedTranslogs.addAll(foundTranslogs);
+        }
+    }
 
-        this.type = FsTranslogFile.Type.fromString(indexSettings.get("index.translog.fs.type", FsTranslogFile.Type.BUFFERED.name()));
-        this.bufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.buffer_size", ByteSizeValue.parseBytesSizeValue("64k")).bytes();
+    /* extracts the translog id from a file name. returns -1 upon failure */
+    public static long parseIdFromFileName(Path translogFile) {
+        final String fileName = translogFile.getFileName().toString();
+        final Matcher matcher = PARSE_ID_PATTERN.matcher(fileName);
+        if (matcher.matches()) {
+            try {
+                return Long.parseLong(matcher.group(1));
+            } catch (NumberFormatException e) {
+                // meh :(
+            }
+        }
+        return -1;
     }
 
     @Override
     public void updateBuffer(ByteSizeValue bufferSize) {
         this.bufferSize = bufferSize.bytesAsInt();
-        rwl.writeLock().lock();
-        try {
-            FsTranslogFile current1 = this.current;
-            if (current1 != null) {
-                current1.updateBufferSize(this.bufferSize);
-            }
-            current1 = this.trans;
-            if (current1 != null) {
-                current1.updateBufferSize(this.bufferSize);
-            }
-        } finally {
-            rwl.writeLock().unlock();
+        try (ReleasableLock lock = writeLock.acquire()) {
+            current.updateBufferSize(this.bufferSize);
         }
     }
 
+    public boolean isOpen() {
```

Comment:
this is usually a bad sign if you need this method? can we have `ensureOpen()` internally instead?